compiler/rustc_parse/src/parser/item.rs

```diff
@@ -2492,6 +2492,19 @@ impl<'a> Parser<'a> {
     pub(super) fn parse_fn_params(&mut self, req_name: ReqName) -> PResult<'a, ThinVec<Param>> {
         let mut first_param = true;
         // Parse the arguments, starting out with `self` being allowed...
+        if self.token.kind != TokenKind::OpenDelim(Delimiter::Parenthesis) {
+            // removed syntax, handled elsewhere
+            if self.token.kind != TokenKind::Tilde
+            // might be typo'd trait impl, handled elsewhere
+            && !self.token.is_keyword(kw::For)
```

Comment:
Could you remove these special cases?
I don't think they are necessary here.

In `fn foo ~` -> `fn foo() ~` and `fn foo for` -> `fn foo() for` the unexpected tokens don't have any special meaning and can be reported equally well after inserting the `()`.

Reply:
There are some specialized diagnostics for `fn Trait for Type` (`impl` -> `fn`) typo and `fn~()` (removed syntax) which don't get reported if we recover and raise the diagnostic here

Reply:
Ok, I see the `for` special case in item.rs, it's silly, but let's keep it.

I don't see any special logic for `~` though and I don't remember such syntax, which means it was removed in 2014 latest, so if any such special treatment exists it is no longer necessary.